### ISSUE TITLE
transformFromAst produces inconsistent sourcemaps

### ISSUE THREAD

**Issue by edoardocavazza** [babel/babel#7046]
Choose one: is this a bug report or feature request?
Bug

### Input Code

```js
const babel = require('@babel/core');

const config = { babelrc: false, sourceMaps: true };

const code = `
class A{
    methodB() {}
}
`;

let result = babel.transform(code, config);
let result2 = babel.transformFromAst(result.ast, code, Object.assign({ inputSourceMap: result.map }, config));
```

### Babel/Babylon Configuration (.babelrc, package.json, cli command)

no .babelrc and no package.json configuration

### Expected Behavior

`babel.transformFromAst` should return the same code and the same mappings of the previous transformation.

### Current Behavior

Generated codes are the same, but mappings are different:

transform: `AACA,MAAMA,CAAN,CAAO;AACHC,YAAU,CAAE;;AADT`
transformFromAst: `AAEIA;aADG;;AACO,AAAE`

### Possible Solution

I really don't know, but if someone could point me where to start, I would be glad to help.

### Context

I would like to run multiple transformations using the AST in order to speed up the build.

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  7.0.0-beta.35
| Babylon          | 
| node             | v9.3.0
| npm              | 5.6.0
| Operating System | macOS 10.13.1

**Comment by edoardocavazza** [comment]
I think the problem is in the sourcemaps merge method [here](https://github.com/babel/babel/blob/e44cef34734b59d26b2f090acd7ab16d5570b05c/packages/babel-core/src/transformation/file/generate.js#L89-L112). Infact, I have noticed that `inputMap` and `map` are equivalent before the merge.

I have tried a different implementation, and it seems to work, but I am not very skilled about sourcemaps generation and I need a review:
```js
outputMapConsumer.eachMapping(function (mapping) {
    var originalPosition = outputMapConsumer.originalPositionFor({
        line: mapping.generatedLine,
        column: mapping.generatedColumn,
        source: source
    });
    if (originalPosition.line !== null) {
        mergedGenerator.addMapping({
            generated: {
                line: mapping.generatedLine,
                column: mapping.generatedColumn
            },
            original: {
                line: originalPosition.line,
                column: originalPosition.column
            },
            source: originalPosition.source,
            name: originalPosition.name
        });
    }
});
```

Basically, instead of searching generated position for the current mapping, this looks for original informations and updates the model.
